### ISSUE TITLE
Body.insertOoxml() crashes Word on IE11 (Windows 10 version 1809)

### ISSUE THREAD

**Issue by DougCoraritoAT** [OfficeDev/office-js#842]
There was a previous bug with body.insertOoxml() that has been fixed when Word launches Edge.  However, when running Windows build 1809, Word launches IE11 and the crash is still present.  Please see this link for more information on what browser Word launches:

https://docs.microsoft.com/en-us/office/dev/add-ins/concepts/browsers-used-by-office-web-add-ins

## Expected Behavior
Calling Body.insertOoxml() should not crash Word when using IE11.

## Current Behavior
Calling Body.insertOoxml() crahses Word when using IE11.

## Steps to Reproduce, or Live Example
name: Blank snippet (5)
description: Create a new snippet from a blank template.
host: WORD
api_set: {}
script:
  content: |
    $("#run").click(() => tryCatch(run));

    async function run() {
      await Word.run(async (context) => {
        const body = context.document.body;
        let bodyOoxmlResult = body.getOoxml();
        await context.sync();

        let bodyOoxml = bodyOoxmlResult.value;
        let parser = new DOMParser();
        let xmlDoc = parser.parseFromString(bodyOoxml, "application/xml");
        let newBodyOoxml = new XMLSerializer().serializeToString(xmlDoc);

        body.clear();
        await context.sync();

        body.insertOoxml(newBodyOoxml, Word.InsertLocation.replace);
        await context.sync();
      });
    }

    /** Default helper for invoking an action and handling errors. */
    async function tryCatch(callback) {
      try {
        await callback();
      } catch (error) {
        // Note: In a production add-in, you'd want to notify the user through your add-in's UI.
        console.error(error);
      }
    }
  language: typescript
template:
  content: |
    <button id="run" class="ms-Button">
        <span class="ms-Button-label">Run</span>
    </button>
  language: html
style:
  content: |-
    section.samples {
        margin-top: 20px;
    }

    section.samples .ms-Button, section.setup .ms-Button {
        display: block;
        margin-bottom: 5px;
        margin-left: 20px;
        min-width: 80px;
    }
  language: css
libraries: |
  https://appsforoffice.microsoft.com/lib/1/hosted/office.js
  @types/office-js

  office-ui-fabric-js@1.4.0/dist/css/fabric.min.css
  office-ui-fabric-js@1.4.0/dist/css/fabric.components.min.css

  core-js@2.4.1/client/core.min.js
  @types/core-js

  jquery@3.1.1
  @types/jquery@3.3.1

Use a blank document and enter sentence "This is a test."

## Context
We are automatically inserting comments and bookmarks.  We would not need to use insertOoxml() if the API allowed us to created/remove comments and bookmarks directly.

## Your Environment
Desktop Word on Windows version 1809 (or any version that launches IE11)

**Comment by tsabbay** [comment]
We are experiencing the same after the last update.
Great job by Microsoft team, again you killed 💥💥💥💥some of the core functionality of office-js.
Seems like you guys NOT checking and the worst part, every fix takes weeks.
Well done!

**Comment by AbidRahman-MSFT** [comment]
Thank you for bringing up this issue. 
@weshi can you provide further context on this issue?

**Comment by weshi** [comment]
@tsabbay while the crash should not happen, can you provide more information why you need below lines instread of inserting the retrieved ooxml from body.getOoxml() directly into the doc? 

    let parser = new DOMParser();
    let xmlDoc = parser.parseFromString(bodyOoxml, "application/xml");
    let newBodyOoxml = new XMLSerializer().serializeToString(xmlDoc);

If you directly use the value of let bodyOoxml = bodyOoxmlResult.value; to call the body.insertOoxml() API, it should succeed without crash.

But anyway we will look into the reason why this caused a crash. Thanks.

**Comment by DougCoraritoAT** [comment]
Hi Wenbo,

To answer your question, we are developing an add-in that adds and removes comments.  Since the current Javascript API does not have functions for manipulating comments, we are adding comments, commentStart, commentEnd and commentRef elements manually.

We get the body ooxml, create a dom, manipulate that structure then serialize back to ooxml and insert.

Please let me know if I can provide more information.

Doug

**Comment by hjkim-MSFT** [comment]
The problem here is that the xml modified by DOMParser is malformed. To work around this problem, you could directly use the value of the return from body.getOoxml() like this.

```
async function run() {
  await Word.run(async (context) => {
    const body = context.document.body;
    let bodyOoxmlResult = body.getOoxml();
    await context.sync();

    let bodyOoxml = bodyOoxmlResult.value;
    body.clear();
    await context.sync();

    body.insertOoxml(bodyOoxml, Word.InsertLocation.replace);
    await context.sync();
  });
}
```

You can also try this by importing this snippet https://gist.github.com/hjkim71/91ace5043c13add529888b3e6b9ac509 into Script Lab (http://aka.ms/getscriptlab).

We're working on a fix for the crash.

**Comment by ghost** [comment]
This issue has been automatically marked as stale because it is marked as needing author feedback but has not had any activity for **4 days**. It will be closed if no further activity occurs **within 3 days of this comment**. Thank you for your interest in Office Add-ins!

**Comment by sureshjoshi** [comment]
For what it's worth, I've run into this problem on Word on MacOS (16.32) for a couple weeks now.  I mentioned it in #106 (which is about OOXML crashes and is 18 months old with no feedback)

**Comment by hjkim-MSFT** [comment]
The crash issue has been fixed. 16.0.12412.10000 or higher builds should contain the fix.

**Comment by juraevuz** [comment]
I am having the same issue with ie11 on windows 10. I extract Ooxml, make some adjustments and when I try to insert it back, Word crashes. The same Office App is working on latest version of Windows 10, which uses Edge to load add-ins. But on older versions with IE11 it crashes.

**Comment by hjkim-MSFT** [comment]
Fixed in 16.0.12412.10000 build.

**Comment by jamesp-mendeley** [comment]
Of course it's fixed in 16.0.12412.10000 - that's [a version of Word that uses EdgeHTML as the runtime](https://docs.microsoft.com/en-us/office/dev/add-ins/concepts/browsers-used-by-office-web-add-ins), and this is specifically an IE11 runtime bug.  Assuming you're on Windows 10 it can only possibly happen in builds prior to 16.0.11629. ;-)

**Comment by jamesp-mendeley** [comment]
### Cause and Potential Workaround

The problem appears to be that while IE11’s DOMParser() implementation faithfully renders the OOXML string returned by body.getOoxml() correctly into a DOM, IE11’s XMLSerializer() implementation has a bad habit of omitting or moving `xmlns:*` attributes from the element they’re declared on in the DOM to the lower-level child element(s) whose tag-names come from these namespaces.
 
For example, in the raw OOXML we get back from body.getOoxml() one of the initial top-level package elements looks like this:
 
```
<?xml version="1.0" standalone="yes"?>
<?mso-application progid="Word.Document"?>
<pkg:package xmlns:pkg="http://schemas.microsoft.com/office/2006/xmlPackage">
    …
    <pkg:part pkg:name="/word/document.xml" pkg:contentType="application/vnd.openxmlformats-officedocument.wordprocessingml.document.main+xml">
        <pkg:xmlData>
            <w:document xmlns:wpc="http://schemas.microsoft.com/office/word/2010/wordprocessingCanvas"
                xmlns:cx="http://schemas.microsoft.com/office/drawing/2014/chartex"
                xmlns:cx1="http://schemas.microsoft.com/office/drawing/2015/9/8/chartex"
                xmlns:cx2="http://schemas.microsoft.com/office/drawing/2015/10/21/chartex"
                xmlns:cx3="http://schemas.microsoft.com/office/drawing/2016/5/9/chartex"
                xmlns:cx4="http://schemas.microsoft.com/office/drawing/2016/5/10/chartex"
                xmlns:cx5="http://schemas.microsoft.com/office/drawing/2016/5/11/chartex"
                xmlns:cx6="http://schemas.microsoft.com/office/drawing/2016/5/12/chartex"
                xmlns:cx7="http://schemas.microsoft.com/office/drawing/2016/5/13/chartex"
                xmlns:cx8="http://schemas.microsoft.com/office/drawing/2016/5/14/chartex"
                xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
                xmlns:aink="http://schemas.microsoft.com/office/drawing/2016/ink"
                xmlns:am3d="http://schemas.microsoft.com/office/drawing/2017/model3d"
                xmlns:o="urn:schemas-microsoft-com:office:office"
                xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships"
                xmlns:m="http://schemas.openxmlformats.org/officeDocument/2006/math"
                xmlns:v="urn:schemas-microsoft-com:vml"
                xmlns:wp14="http://schemas.microsoft.com/office/word/2010/wordprocessingDrawing"
                xmlns:wp="http://schemas.openxmlformats.org/drawingml/2006/wordprocessingDrawing"
                xmlns:w10="urn:schemas-microsoft-com:office:word"
                xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main"
                xmlns:w14="http://schemas.microsoft.com/office/word/2010/wordml"
                xmlns:w15="http://schemas.microsoft.com/office/word/2012/wordml"
                xmlns:w16cid="http://schemas.microsoft.com/office/word/2016/wordml/cid"
                xmlns:w16se="http://schemas.microsoft.com/office/word/2015/wordml/symex"
                xmlns:wpg="http://schemas.microsoft.com/office/word/2010/wordprocessingGroup"
                xmlns:wpi="http://schemas.microsoft.com/office/word/2010/wordprocessingInk"
                xmlns:wne="http://schemas.microsoft.com/office/word/2006/wordml"
                xmlns:wps="http://schemas.microsoft.com/office/word/2010/wordprocessingShape" mc:Ignorable="w14 w15 w16se w16cid wp14">
                <w:body>
                …
```
 
As you can see there are a large number of XML namespaces defined on the pkg:package  > pkg:part> pkg>xmlData > w:document element, and querying the resulting DOM object shows that these are faithfully retained rendered into the appropriate element in the DOM constructed by DOMParser().
 
When you call XMLSerializer() to render the DOM back into an OOXML string though, it returns something more like the following:
 
```
<pkg:package xmlns:pkg="http://schemas.microsoft.com/office/2006/xmlPackage">
    …
    <pkg:part pkg:contentType="application/vnd.openxmlformats-officedocument.wordprocessingml.document.main+xml" pkg:name="/word/document.xml">
        <pkg:xmlData>
            <w:document xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main"
                xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" mc:Ignorable="w14 w15 w16se w16cid wp14">
                <w:body>
                …
 ```

… where the missing namespaces have been omitted (if not actually used anywhere in the OOXML tree), or have been moved lower in the tree, to the first instances of actual elements that come from those namespaces.
 
Manually replacing the missing XML processing directives at the beginning Using a third-party XML serializer from npm (eg, [xmlserializer](https://www.npmjs.com/package/xmlserializer)) that correctly leaves the `xmlns:*` attributes where they are in the DOM (instead of trying to get cute and having opinions on which ones are actually needed and where) seems to fix the crashing bug that takes down the entirety of Word.
